File: common/src/main/java/mconi/common/sim/OniSimulationGrid.java
package mconi.common.sim;

import mconi.common.sim.model.OniCellCoordinate;
import mconi.common.sim.model.OniCellState;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Active-cell storage for the simulation kernel.
 */
public class OniSimulationGrid
{
	private final Map<OniCellCoordinate, OniCellState> cells = new ConcurrentHashMap<>();

	public OniCellState getOrCreateCellAtBlock(int x, int y, int z, int cellSize)
	{
		OniCellCoordinate coordinate = OniCellCoordinate.fromBlockPosition(x, y, z, cellSize);
		return cells.computeIfAbsent(coordinate, ignored -> new OniCellState());
	}

	public int activeCellCount()
	{
		return cells.size();
	}

	public void clear()
	{
		cells.clear();
	}
}

File: common/src/main/kotlin/mconi/common/sim/OniSimulationGrid.kt
package mconi.common.sim

import mconi.common.sim.model.OniCellCoordinate
import mconi.common.sim.model.OniCellState
import java.util.concurrent.ConcurrentHashMap

/**
 * Active-cell storage for the simulation kernel.
 */
class OniSimulationGrid {
    private val cells: MutableMap<OniCellCoordinate, OniCellState> = ConcurrentHashMap()

    fun getOrCreateCellAtBlock(x: Int, y: Int, z: Int, cellSize: Int): OniCellState {
        val coordinate = OniCellCoordinate.fromBlockPosition(x, y, z, cellSize)
        return cells.computeIfAbsent(coordinate) { OniCellState() }
    }

    fun getOrCreateCellAtCoordinate(coordinate: OniCellCoordinate): OniCellState {
        return cells.computeIfAbsent(coordinate) { OniCellState() }
    }

    fun getCellAtCoordinate(coordinate: OniCellCoordinate): OniCellState? = cells[coordinate]

    fun cellEntries(): Set<Map.Entry<OniCellCoordinate, OniCellState>> = cells.entries

    fun activeCellCount(): Int = cells.size

    fun cells(): Collection<OniCellState> = cells.values

    fun clear() {
        cells.clear()
    }
}
